art/attacks/evasion/imperceptible_asr/imperceptible_asr.py

```diff
@@ -0,0 +1,621 @@
+        :param psd_maximum_stabilized: Stabilized maximum across frames, i.e. shape is `(batch_size, frame_length)`, of
+            the original unnormalized PSD of `x`.
+        :param masking_threshold_stabilized: Stabilized masking threshold for the original input `x`.
+        :return: Approximate PSD tensor of shape `(batch_size, window_size // 2 + 1, frame_length)`.
+        """
+        import tensorflow.compat.v1 as tf1
+
+        # calculate approximate power spectral density
+        psd_perturbation = self._approximate_power_spectral_density_tf(perturbation, psd_maximum_stabilized)
+
+        # calculate hinge loss
+        loss = tf1.reduce_mean(
+            tf1.nn.relu(psd_perturbation - masking_threshold_stabilized), axis=[1, 2], keepdims=False
+        )
+
+        # compute loss gradient
+        loss_gradient = tf1.gradients(loss, [perturbation])[0]
+        return loss_gradient, loss
+
+    def _approximate_power_spectral_density_tf(
+        self, perturbation: "Tensor", psd_maximum_stabilized: "Tensor"
+    ) -> "Tensor":
+        """
+        Approximate the power spectral density for a perturbation `perturbation` in TensorFlow.
+
+        Note that a stabilized PSD approximate is returned, where the `10*log10`-term has been canceled out.
+        Following Qin et al (2019) this mitigates optimization instabilities.
+
+        :param perturbation: Adversarial perturbation.
+        :param psd_maximum_stabilized: Stabilized maximum across frames, i.e. shape is `(batch_size, frame_length)`, of
+            the original unnormalized PSD of `x`.
+        :return: Approximate PSD tensor of shape `(batch_size, window_size // 2 + 1, frame_length)`.
+        """
+        import tensorflow.compat.v1 as tf1
+
+        # compute short-time Fourier transform (STFT)
+        stft_matrix = tf1.signal.stft(perturbation, self._window_size, self._hop_size, fft_length=self._window_size)
+
+        # compute power spectral density (PSD)
+        # note: fixes implementation of Qin et al. by also considering the square root of gain_factor
+        gain_factor = 8.0 / 3.0
+        psd_matrix = gain_factor * tf1.square(tf1.abs(stft_matrix / self._window_size))
+
+        # approximate normalized psd: psd_matrix_approximated = 10^((96.0 - psd_matrix_max + psd_matrix)/10)
+        psd_matrix_approximated = tf1.pow(10.0, 9.6) / tf1.expand_dims(psd_maximum_stabilized, -1) * psd_matrix
+
+        # return PSD matrix such that shape is (batch_size, window_size // 2 + 1, frame_length)
+        return tf1.transpose(psd_matrix_approximated, [0, 2, 1])
+
+    def _approximate_power_spectral_density_torch(self):
+        """Approximate the power spectral density for a perturbation `perturbation` in PyTorch."""
+        raise NotImplementedError
+
+    def _loss_gradient_masking_threshold_torch(self):
+        """Compute loss gradient of the masking threshold loss in PyTorch."""
+        raise NotImplementedError
```

Comment:
What is the reason for these two methods?

Reply:
Very soon these will be used to add support for any PyTorch ASR estimator :100: